### ISSUE TITLE
Windows: pkg.latest state not updating packages.

### ISSUE THREAD

**Issue by whytewolf** [saltstack/salt#47484]
### Description of Issue/Question
This affects both py2 and py3

Installed Firefox 42.0, and used a state with pkg.latest to attempt to upgrade to the latest release. 

state returns 

```
winion-1:
----------
          ID: update-firefox
    Function: pkg.latest
        Name: firefox
      Result: False
     Comment: Package firefox failed to update.
     Started: 21:17:51.152000
    Duration: 19071.0 ms
     Changes:

Summary for winion-1
------------
Succeeded: 0
Failed:    1
------------
Total states run:     1
Total run time:  19.071 s
```

### Setup

Windows minion running 2018.3.0 with a state with pkg.latest and a package that needs updating. 

```
[root@master salt-winrepo-ng]# salt -G 'os:windows' pkg.list_pkgs
winion-1:
    ----------
    Microsoft Silverlight:
        5.1.50907.0
    Mozilla Maintenance Service:
        42.0
    Oracle VM VirtualBox Guest Additions 5.1.22:
        5.1.22.0
    firefox:
        42.0
    npp:
        7.5.6
    salt-minion:
        2018.3.0
```

```
[root@master salt-winrepo-ng]# salt -G 'os:windows' pkg.latest_version firefox
winion-1:
    59.0.3
```

```
[root@master salt-winrepo-ng]# cat /srv/salt/update.sls
update-firefox:
  pkg.latest:
    - name: firefox
```

### Steps to Reproduce Issue
See section above for setup. 

in the debug log line was all the rendering of the win_repo pacakges
afterwords was this 

```
[DEBUG   ] Latest installed version of firefox is 42.0
[DEBUG   ] Latest available version of package firefox is 59.0.3
[DEBUG   ] Upgrade of firefox from 42.0 to 59.0.3 is available
[DEBUG   ] Using existing pkg metadata db for saltenv 'base' (age is 0:00:00.047126)
[DEBUG   ] Could not LazyLoad pkg.normalize_name: 'pkg.normalize_name' is not available.
[DEBUG   ] Using existing pkg metadata db for saltenv 'base' (age is 0:00:00.062777)
[DEBUG   ] A version (None) already installed for package firefox
[DEBUG   ] Using existing pkg metadata db for saltenv 'base' (age is 0:00:00.078435)
[ERROR   ] Package firefox failed to update.
[INFO    ] Completed state [firefox] at time 21:10:23.051718 (duration_in_ms=16382.652)
```

After attempting:
```
winion-1:
    ----------
    Microsoft Silverlight:
        5.1.50907.0
    Mozilla Maintenance Service:
        42.0
    Oracle VM VirtualBox Guest Additions 5.1.22:
        5.1.22.0
    firefox:
        42.0
    npp:
        7.5.6
    salt-minion:
        2018.3.0
```

### Versions Report

```
winion-0:
    Salt Version:
               Salt: 2018.3.0

    Dependency Versions:
               cffi: 1.10.0
           cherrypy: 10.2.1
           dateutil: 2.6.1
          docker-py: Not Installed
              gitdb: 2.0.3
          gitpython: Not Installed
              ioflo: Not Installed
             Jinja2: 2.10
            libgit2: Not Installed
            libnacl: Not Installed
           M2Crypto: Not Installed
               Mako: 1.0.7
       msgpack-pure: Not Installed
     msgpack-python: 0.4.8
       mysql-python: Not Installed
          pycparser: 2.18
           pycrypto: 2.6.1
       pycryptodome: Not Installed
             pygit2: Not Installed
             Python: 3.5.3 (v3.5.3:1880cb95a742, Jan 16 2017, 16:02:32) [MSC v.1900 64 bit (AMD64)]
       python-gnupg: 0.4.1
             PyYAML: 3.12
              PyZMQ: 16.0.3
               RAET: Not Installed
              smmap: 2.0.3
            timelib: 0.2.4
            Tornado: 4.5.2
                ZMQ: 4.1.6

    System Versions:
               dist:
             locale: cp1252
            machine: AMD64
            release: 2016Server
             system: Windows
            version: 2016Server 10.0.14393 SP0 Multiprocessor Free
```

```
winion-1:
    Salt Version:
               Salt: 2018.3.0

    Dependency Versions:
               cffi: 1.10.0
           cherrypy: 10.2.1
           dateutil: 2.6.1
          docker-py: Not Installed
              gitdb: 2.0.3
          gitpython: Not Installed
              ioflo: Not Installed
             Jinja2: 2.10
            libgit2: Not Installed
            libnacl: 1.6.1
           M2Crypto: Not Installed
               Mako: 1.0.7
       msgpack-pure: Not Installed
     msgpack-python: 0.4.8
       mysql-python: Not Installed
          pycparser: 2.18
           pycrypto: 2.6.1
       pycryptodome: Not Installed
             pygit2: Not Installed
             Python: 2.7.14 (v2.7.14:84471935ed, Sep 16 2017, 20:25:58) [MSC v.1500 64 bit (AMD64)]
       python-gnupg: 0.4.1
             PyYAML: 3.12
              PyZMQ: 16.0.3
               RAET: Not Installed
              smmap: 2.0.3
            timelib: 0.2.4
            Tornado: 4.5.2
                ZMQ: 4.1.6

    System Versions:
               dist:
             locale: cp1252
            machine: AMD64
            release: 2016Server
             system: Windows
            version: 2016Server 10.0.14393  Multiprocessor Free
```

**Comment by whytewolf** [comment]
ZD-2495

**Comment by damon-atkins** [comment]
Please provide the sls file used for package definition.

**Comment by whytewolf** [comment]
Sure not a problem. here are the three I just tried
* [firefox](https://github.com/saltstack/salt-winrepo-ng/blob/master/firefox.sls)
* [keepass](https://github.com/saltstack/salt-winrepo-ng/blob/master/keepass.sls)
* [npp](https://github.com/saltstack/salt-winrepo-ng/blob/master/npp.sls)

```
salt -G 'os:windows' pkg.list_pkgs
winion-0:
    ----------
    Microsoft Silverlight:
        5.1.50907.0
    Mozilla Maintenance Service:
        42.0
    Oracle VM VirtualBox Guest Additions 5.1.22:
        5.1.22.0
    firefox:
        42.0
    keepass:
        1.34.0
    npp:
        7.4.2
    salt-minion:
        2018.3.0
```

The last lines after the rendering of win-repo-ng for pkg.latest_version firefire
```
[DEBUG   ] LazyLoaded reg.list_keys
[DEBUG   ] Latest installed version of firefox is 42.0
[DEBUG   ] Latest available version of package firefox is 59.0.3
[DEBUG   ] Upgrade of firefox from 42.0 to 59.0.3 is available
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for (u'c:\\salt\\conf\\pki\\minion', u'winion-0',
u'tcp://192.168.70.10:4506', u'aes')
[DEBUG   ] Initializing new AsyncAuth for (u'c:\\salt\\conf\\pki\\minion', u'winion-0', u'tcp://192.168.70.10:4506')
[DEBUG   ] Connecting the Minion to the Master URI (for the return server): tcp://192.168.70.10:4506
[DEBUG   ] Trying to connect to: tcp://192.168.70.10:4506
[DEBUG   ] LazyLoaded nested.output
local:
    59.0.3
```
Same for keepass
```
[DEBUG   ] LazyLoaded reg.list_keys
[DEBUG   ] Latest installed version of keepass is 1.34.0
[DEBUG   ] Latest available version of package keepass is 1.35.0
[DEBUG   ] Upgrade of keepass from 1.34.0 to 1.35.0 is available
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for (u'c:\\salt\\conf\\pki\\minion', u'winion-0',
u'tcp://192.168.70.10:4506', u'aes')
[DEBUG   ] Initializing new AsyncAuth for (u'c:\\salt\\conf\\pki\\minion', u'winion-0', u'tcp://192.168.70.10:4506')
[DEBUG   ] Connecting the Minion to the Master URI (for the return server): tcp://192.168.70.10:4506
[DEBUG   ] Trying to connect to: tcp://192.168.70.10:4506
[DEBUG   ] LazyLoaded nested.output
local:
    1.35.0
```
and npp
```
[DEBUG   ] LazyLoaded reg.list_keys
[DEBUG   ] Latest installed version of npp is 7.4.2
[DEBUG   ] Latest available version of package npp is 7.5.6
[DEBUG   ] Upgrade of npp from 7.4.2 to 7.5.6 is available
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for (u'c:\\salt\\conf\\pki\\minion', u'winion-0',
u'tcp://192.168.70.10:4506', u'aes')
[DEBUG   ] Initializing new AsyncAuth for (u'c:\\salt\\conf\\pki\\minion', u'winion-0', u'tcp://192.168.70.10:4506')
[DEBUG   ] Connecting the Minion to the Master URI (for the return server): tcp://192.168.70.10:4506
[DEBUG   ] Trying to connect to: tcp://192.168.70.10:4506
[DEBUG   ] LazyLoaded nested.output
local:
    7.5.6
```

Here is the full debug from trying to pkg.install the firefox package that wants an update, like [pkg.latest](https://github.com/saltstack/salt/blob/develop/salt/states/pkg.py#L2470-L2475) runs it.
```
salt-call : [DEBUG   ] Reading configuration from c:\salt\conf\minion
At line:1 char:1
+ salt-call -l debug pkg.install firefox *> C:\vagrant\firefox.install_ ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: ([DEBUG   ] Read...alt\conf\minion:String) [], RemoteException
    + FullyQualifiedErrorId : NativeCommandError

[DEBUG   ] Including configuration from 'c:\salt\conf\minion.d\_schedule.conf'
[DEBUG   ] Reading configuration from c:\salt\conf\minion.d\_schedule.conf
[DEBUG   ] Using cached minion ID from c:\salt\conf\minion_id: winion-0
[DEBUG   ] Configuration file path: c:\salt\conf\minion
[WARNING ] Insecure logging configuration detected! Sensitive data may be logged.
[DEBUG   ] Grains refresh requested. Refreshing grains.
[DEBUG   ] Reading configuration from c:\salt\conf\minion
[DEBUG   ] Including configuration from 'c:\salt\conf\minion.d\_schedule.conf'
[DEBUG   ] Reading configuration from c:\salt\conf\minion.d\_schedule.conf
[DEBUG   ] Connecting to master. Attempt 1 (infinite attempts)
[DEBUG   ] Master URI: tcp://192.168.70.10:4506
[DEBUG   ] Popen(['git', 'version'], cwd=C:\salt, universal_newlines=False, shell=None)
[DEBUG   ] Initializing new AsyncAuth for (u'c:\\salt\\conf\\pki\\minion', u'winion-0', u'tcp://192.168.70.10:4506')
[DEBUG   ] Generated random reconnect delay between '1000ms' and '11000ms' (4973)
[DEBUG   ] Setting zmq_reconnect_ivl to '4973ms'
[DEBUG   ] Setting zmq_reconnect_ivl_max to '11000ms'
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for (u'c:\\salt\\conf\\pki\\minion', u'winion-0', 
u'tcp://192.168.70.10:4506', 'clear')
[DEBUG   ] Connecting the Minion to the Master URI (for the return server): tcp://192.168.70.10:4506
[DEBUG   ] Trying to connect to: tcp://192.168.70.10:4506
[DEBUG   ] salt.crypt.get_rsa_pub_key: Loading public key
[DEBUG   ] Decrypting the current master AES key
[DEBUG   ] salt.crypt.get_rsa_key: Loading private key
[DEBUG   ] salt.crypt._get_key_with_evict: Loading private key
[DEBUG   ] Loaded minion key: c:\salt\conf\pki\minion\minion.pem
[DEBUG   ] salt.crypt.get_rsa_pub_key: Loading public key
[DEBUG   ] Connecting the Minion to the Master publish port, using the URI: tcp://192.168.70.10:4505
[DEBUG   ] salt.crypt.get_rsa_key: Loading private key
[DEBUG   ] Loaded minion key: c:\salt\conf\pki\minion\minion.pem
[DEBUG   ] Determining pillar cache
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for (u'c:\\salt\\conf\\pki\\minion', u'winion-0', 
u'tcp://192.168.70.10:4506', u'aes')
[DEBUG   ] Initializing new AsyncAuth for (u'c:\\salt\\conf\\pki\\minion', u'winion-0', u'tcp://192.168.70.10:4506')
[DEBUG   ] Connecting the Minion to the Master URI (for the return server): tcp://192.168.70.10:4506
[DEBUG   ] Trying to connect to: tcp://192.168.70.10:4506
[DEBUG   ] salt.crypt.get_rsa_key: Loading private key
[DEBUG   ] Loaded minion key: c:\salt\conf\pki\minion\minion.pem
[DEBUG   ] LazyLoaded jinja.render
[DEBUG   ] LazyLoaded yaml.render
[DEBUG   ] LazyLoaded pkg.install
[DEBUG   ] SMBIOS: neither dmidecode nor smbios found!
[WARNING ] C:\salt\bin\lib\site-packages\salt\modules\win_update.py:92: DeprecationWarning: The 'win_update' module is 
being deprecated and will be removed in Salt Fluorine (Unreleased). Please use the 'win_wua' module instead.

[DEBUG   ] Marking 'strftime' as a jinja filter
[DEBUG   ] Could not LazyLoad pkg.normalize_name: 'pkg.normalize_name' is not available.
[DEBUG   ] Using existing pkg metadata db for saltenv 'base' (age is 0:00:37.559578)
[DEBUG   ] A version (None) already installed for package firefox
[DEBUG   ] Using existing pkg metadata db for saltenv 'base' (age is 0:00:37.606578)
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for (u'c:\\salt\\conf\\pki\\minion', u'winion-0', 
u'tcp://192.168.70.10:4506', u'aes')
[DEBUG   ] Initializing new AsyncAuth for (u'c:\\salt\\conf\\pki\\minion', u'winion-0', u'tcp://192.168.70.10:4506')
[DEBUG   ] Connecting the Minion to the Master URI (for the return server): tcp://192.168.70.10:4506
[DEBUG   ] Trying to connect to: tcp://192.168.70.10:4506
[DEBUG   ] LazyLoaded nested.output
local:
    ----------
```

**Comment by damon-atkins** [comment]
pkg.install has done the correct action above, the software (any version) is already installed, therefore no action required.

Please use the following for debugging.
`salt-call -l debug pkg.install firefox version=59.0.3`
OR
`salt-call -l debug pkg.latest firefox`

**Comment by damon-atkins** [comment]
The change was documented
From https://github.com/saltstack/salt/blob/develop/doc/topics/releases/2017.7.3.rst &
From https://github.com/saltstack/salt/blob/develop/doc/topics/releases/2016.11.9.rst
`pkg.install without a version parameter no longer upgrades software if the software is already installed. Use pkg.install version=latest or in a state use pkg.latest to get the old behavior.`

I am surprised this was not picked up in the 20+ tests.  Need to review the tests.

Could you please provide the package sls file, your using please.

The fix for pkg.latest **state** is to call pkg.install with
`pkgs=[{"foo": "latest"}, {"bar": "latest"}] ` or `pkgs=[{"foo": "<latest version>"}, {"bar": "<latest version>"}]`

The state goes through the effort of compating versions to determine if a change is required however, it does not use the version information when it calls pkg.install.

**Comment by damon-atkins** [comment]
This is the required fix for pkg.latest state. This needs to be applied to 2016.11 first.
salt.utils.platform.is_windows() will need to be change for the specific version of salt.
Not sure why there is a need for an `if pkgs`  test in the original code.

```
                    'comment': '\n'.join(comments)}
        ## Remove Next Two Lines
        ## Build updated list of pkgs to exclude non-targeted ones
        #targeted_pkgs = [{x: targets[x]} for x in targets.keys()] if pkgs else None

        try:
            # No need to refresh, if a refresh was necessary it would have been
            # performed above when pkg.latest_version was run.
            if salt.utils.platform.is_windows():
                # Build updated list of pkgs to exclude non-targeted ones
                targeted_pkgs = [{x: targets[x]} for x in targets.keys()]
                changes = __salt__['pkg.install'](name=None,
                                                  refresh=False,
                                                  fromrepo=fromrepo,
                                                  skip_verify=skip_verify,
                                                  pkgs=targeted_pkgs,
                                                  **kwargs)
            else:
                # Build updated list of pkgs to exclude non-targeted ones
                targeted_pkgs = [{x: targets[x]} for x in targets.keys()] if pkgs else None
                changes = __salt__['pkg.install'](name,
                                                  refresh=False,
                                                  fromrepo=fromrepo,
                                                  skip_verify=skip_verify,
                                                  pkgs=targeted_pkgs,
                                                  **kwargs)
        except CommandExecutionError as exc:
```

**Comment by rallytime** [comment]
@whytewolf The fix from @damon-atkins will be in the 2017.7.6 and 2018.3.1 releases. Can you confirm that the issue is fixed for you?

**Comment by twangboy** [comment]
I can confirm this issue is fixed on tag v2017.7.6
